rust/noosphere-core/src/data/body_chunk.rs

```diff
@@ -70,4 +76,342 @@ impl BodyChunkIpld {
+        }
+    }
+
+    pub async fn push(&mut self, chunk: Vec<u8>) -> Result<()> {
+        match self {
+            BlobVecData::Memory {
+                chunks,
+                ref mut chunk_count,
+                ref mut byte_length,
+            } => {
+                *byte_length += chunk.len() as u64;
+                chunks.push(chunk);
+                *chunk_count += 1;
+            }
+            BlobVecData::Db {
+                store,
+                ref mut chunk_count,
+                ref mut byte_length,
+                ..
+            } => {
+                *byte_length += chunk.len() as u64;
+                let key = format!("{}", chunk_count);
+                store.set_key(&key, chunk).await?;
+                *chunk_count += 1;
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn into_stream(
+        self,
+        is_reversed: bool,
+    ) -> impl Stream<Item = Result<Vec<u8>, anyhow::Error>> + Unpin {
+        Box::pin(try_stream! {
+        match self {
+            BlobVecData::Memory {
+                chunks,
+                ..
+            } => {
+                if is_reversed {
+                    let iterator = chunks.into_iter().rev();
+                    for chunk in iterator {
+                        yield chunk;
+                    }
+                } else {
+                    let iterator = chunks.into_iter();
+                    for chunk in iterator {
+                        yield chunk;
+                    }
+                }
+            },
+            BlobVecData::Db{
+                store,
+                chunk_count,
+                ..
+            } => {
+                for n in 0..chunk_count {
+                    let index = if is_reversed {
+                        chunk_count - n - 1
+                    } else {
+                        n
+                    };
+                    let key = format!("{}", index);
+                    let chunk = store.get_key(&key).await?;
+                    yield chunk.ok_or_else(|| anyhow!("Missing chunk."))?;
+                }
+            }
+        }
+        })
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use noosphere_storage::{helpers::make_disposable_storage, MemoryStore, SphereDb};
+    use tokio_stream::StreamExt;
+
+    #[cfg(target_arch = "wasm32")]
+    use wasm_bindgen_test::wasm_bindgen_test;
+    #[cfg(target_arch = "wasm32")]
+    wasm_bindgen_test::wasm_bindgen_test_configure!(run_in_browser);
+
+    #[cfg_attr(target_arch = "wasm32", wasm_bindgen_test)]
+    #[cfg_attr(not(target_arch = "wasm32"), tokio::test)]
+    async fn it_reads_and_writes_chunks_nonstreaming() -> Result<()> {
+        let mut store = MemoryStore::default();
+        let chunk1 = vec![1; BODY_CHUNK_MAX_SIZE.try_into().unwrap()];
+        let chunk2 = vec![2; BODY_CHUNK_MAX_SIZE.try_into().unwrap()];
+        let chunk3 = vec![3; <u32 as TryInto::<usize>>::try_into(BODY_CHUNK_MAX_SIZE).unwrap() / 2];
```

Comment:
It would be worth filling this test data with non-zero values, since a slice of zeros of the same size will get the same CID (and this could lend itself to a misleading outcome in the test).

Reply:
The `vec![1; SIZE]` format uses the first value as an initializer, so these chunks are `[1, 1, 1, 1, ..]`, `[2, 2, 2, ..]` etc